### ISSUE TITLE
Using wrong XSD type for URIorCURIE

### ISSUE THREAD

**Issue by Silvanoc** [linkml/linkml#2213]
**Describe the bug**

[Urieorcurie type](https://linkml.io/linkml-model/latest/docs/Uriorcurie/) is declared with URI [xsd:anyURI](http://www.w3.org/2001/XMLSchema#anyURI), what is wrong since a CURIE might not be a valid URI.

**To reproduce**
1. Get an XML Schema Validator (I have tested it with [this](https://www.liquid-technologies.com/online-xsd-validator) and [this](https://www.freeformatter.com/xml-validator-xsd.html) online validators )
2. Provide an XML Schema that specifies the type `xsd:anyURI` on an attribute value. This is the one that I have used: ```<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema">
  <xs:element name="slot_type">
    <xs:complexType>
      <xs:attribute name="src" type="xs:anyURI"/>
    </xs:complexType>
  </xs:element>
</xs:schema>```
3. Provide a minimalistic XML document containing a CURIE with an underscore `'_'` in the prefix. This is what I have used: `<slot_type src="pre_fix:reference"/>`
4. See error `cvc-datatype-valid.1.2.1: 'pre_fix:reference' is not a valid value for 'anyURI'.` or similar.

**Expected behavior**
A type that gets a XML-Schema URI as its URI, should comply with the corresponding XML-Schema.

The is no type in the ["W3C XML Schema Definition Language (XSD) 1.1"](https://www.w3.org/TR/xmlschema11-2) for CURIEs. The CURIE specification provides an [XML-Schema for CURIEs](https://www.w3.org/TR/2010/NOTE-curie-20101216/#sec_A.1.), but this does not help for the URI of the LinkML `Uriorcurie` or `Curie` types.

In this specific case, having `xsd:anyURI` as URI for `Uriorcurie` LinkML type should mean that any value that is a valid URI or CURIE should pass the XML-Schema validation, what is not true for the valid CURIE [`test_pref:Boat`](https://github.com/linkml/linkml/blob/669fbff40a2e7683797d0c4e510bc21dee7815aa/tests/test_issues/input/linkml_issue_1608_data.yaml#L4).

**Additional context**
LinkML Model version: 1.8.x

I have discovered the issue by chance when trying to improve the JSON-Schema generated for URI, CURIE and URIorCURIE in MR #2212.

In commit 82b753b1ae9cf23b06707b04cb8838dea654b161 I have assigned JSON-Schema type `string` with format `uri` for all there types, assuming that CURIEs would be valid URIs because of the use of `xsd:anyURI` for the URI of the LinkML type `Uriorcurie`.

Luckily [a test](https://github.com/linkml/linkml/blob/main/tests/test_issues/test_linkml_issue_1608.py) exists covering the corner case of a CURIE prefix containing an underscore: [`test_pref:Boat`](https://github.com/linkml/linkml/blob/669fbff40a2e7683797d0c4e510bc21dee7815aa/tests/test_issues/input/linkml_issue_1608_data.yaml#L4) and [this test is failing](https://github.com/linkml/linkml/actions/runs/9995680734/job/27628402136#step:9:446) with the changes proposed in the PR #2212.

**Comment by mahdanoura** [comment]
> what is wrong since a CURIE might not be a valid URI.
Yes, this is clearly stated in the [CURIE syntax Note document](https://www.w3.org/TR/2010/NOTE-curie-20101216/#sec_2.2.): 

`[CURIEs](https://www.w3.org/TR/2010/NOTE-curie-20101216/#P_curie) and [SafeCURIEs](https://www.w3.org/TR/2010/NOTE-curie-20101216/#P_safe_curie) map to IRIs, but neither a CURIE nor a Safe_CURIE is an IRI or URI.`

**Comment by turbomam** [comment]
Thanks @Silvanoc  and @mahdanoura. You guys are much better at interpreting and testing against W3C specifications than I am. I would still like to add something to this issue:

The National Microbiome Data Collaborative has as use case in which the values of a `uriorcurie`-type slot are expected to be converted into CURIes or URIs as part of `linkml-convert` conversion form JSON or YAML into RDF. This doesn't currently work. As you might imagine, the values are just asserted as `xsd:anyURI`-typed string literals.

I have written [some Python code](https://github.com/microbiomedata/nmdc-schema/blob/main/nmdc_schema/anyuri_strings_to_iris.py) that detects the `xsd:anyURI`-typed strings in a Turtle serialization and does the conversion to CURIes. @cmungall is aware of this and I think he has some intention of addressing it in LinkML. In the mean time, I hope we can keep the association between [Urieorcurie type](https://linkml.io/linkml-model/latest/docs/Uriorcurie/) and [xsd:anyURI](http://www.w3.org/2001/XMLSchema#anyURI), becasue without it, I don't see how I make the last step conversion to Turtle CURIes.

We don't have any [prefixes containing underscores](https://github.com/microbiomedata/nmdc-schema/blob/main/src/schema/nmdc.yaml#L32) in our main schema file. I don't think we have any in the other import either but I haven't checked yet.

**Comment by Silvanoc** [comment]
> The National Microbiome Data Collaborative has as use case in which the values of a uriorcurie-type slot are expected to be converted into CURIes or URIs as part of linkml-convert conversion form JSON or YAML into RDF. This doesn't currently work. As you might imagine, the values are just asserted as xsd:anyURI-typed string literals.

The `Uriorcurie` type should accept both, therefore any change fulfilling this requirement should not break this use-case.

WRT the assertion as `xsd:anyURI`, I wonder how/where that assertion is taking place. AFAIK no XML-Schema validation is taking place and most part of the validation takes place checking against a generated JSON-Schema. In that sense the usage of `xsd:anyURI` to specify the URI of the type `Uriorcurie` is only a hint for the JSON-Schema generator to know how to validate it. My expectation would be that any change in the JSON-Schema generation that keeps the validity of the value either as an URI or as a CURIE should not break anything.

In the end IMO as long as we are confident that the tests are covering that use-case, we should be able to work on improving the validation.

> We don't have any [prefixes containing underscores](https://github.com/microbiomedata/nmdc-schema/blob/main/src/schema/nmdc.yaml#L32) in our main schema file. I don't think we have any in the other import either but I haven't checked yet.

You might not have any, but:

1. For whatever reason, you have an example containing one: https://github.com/linkml/linkml/blob/669fbff40a2e7683797d0c4e510bc21dee7815aa/tests/test_issues/input/linkml_issue_1608_data.yaml#L4
2. For LinkML to comply with the W3C specification of the CURIE format, it must support prefixes containing underscores. And IMO LinkML should comply with the standards it claims to support.

The current situation is that you are declaring an XML-Schema type that does not comply with the specification (an XML-Schema validation of `test_pref:Boat` where an `Uriorcurie` is expected would fail). But since the validation is using a JSON-Schema type much more relaxed than the declared XML-Schema, nobody seems to notice it.

**Comment by turbomam** [comment]
OMG, I **did** write that test. I agree that it was a bad choice and am in support of any mechanism that would invalidate it!

**Comment by Silvanoc** [comment]
> OMG, I **did** write that test. I agree that it was a bad choice and am in support of any mechanism that would invalidate it!

Hopefully you don't mean to invalidate the test... Because that test is the very single one ensuring that we have at least one for CURIEs that are valid, but the can not be mistaken for an URI just because they are syntactically valid.

**Comment by Silvanoc** [comment]
> I have written [some Python code](https://github.com/microbiomedata/nmdc-schema/blob/main/nmdc_schema/anyuri_strings_to_iris.py) that detects the `xsd:anyURI`-typed strings in a Turtle serialization and does the conversion to CURIes. @cmungall is aware of this and I think he has some intention of addressing it in LinkML. In the mean time, I hope we can keep the association between [Urieorcurie type](https://linkml.io/linkml-model/latest/docs/Uriorcurie/) and [xsd:anyURI](http://www.w3.org/2001/XMLSchema#anyURI), becasue without it, I don't see how I make the last step conversion to Turtle CURIes.

@turbomam  how do you handle [in your code](https://github.com/microbiomedata/nmdc-schema/blob/main/nmdc_schema/anyuri_strings_to_iris.py) references of LinkML type `Curie`? Because they [are going to get the datatype `xsd:string`](https://github.com/linkml/linkml-model/blob/76bc3279981f00f9f6590caf7fbe5a1f15520f85/linkml_model/model/schema/types.yaml#L134), right? How do you want to make them apart from simple strings? Since [strings also get `xsd:string`](https://github.com/linkml/linkml-model/blob/76bc3279981f00f9f6590caf7fbe5a1f15520f85/linkml_model/model/schema/types.yaml#L30).

Apart from that, since both LinkML types `Uri` and [`UriOrCurie`](https://github.com/linkml/linkml-model/blob/76bc3279981f00f9f6590caf7fbe5a1f15520f85/linkml_model/model/schema/types.yaml#L126) are getting the URI `xsd:anyURI`, you need to "probe" both, although it should be needed only for `UriOrCurie`.

https://github.com/linkml/linkml-model/pull/202 (base for discussion as of now) is trying to resolve those ambiguities, you might want to have a look at it. The fact that `UriOrCurie` has [some space for ambiguity](https://www.w3.org/TR/2010/NOTE-curie-20101216/#sec_2.2.) is known and intrinsically accepted in this type, since we are not using [SafeCURIEs](https://www.w3.org/TR/2010/NOTE-curie-20101216/#P_safe_curie). But at least my proposal constraint those ambiguities to that type.

**Comment by sierra-moxon** [comment]
from dev call:
let's focus on fixing the RDFWriter to convert this correctly.
